File: cpp/remove_element.cc
/*
 * Filename: remove_element.cc
 * Author: Jun Xu
 * Email: jxu.vimacs@gmail.com
 * Created Time: Fri 24 Jul 2015 02:28:12 PM EDT
 */

#include <iostream>
using namespace std;

class Solution {
	public:
		int removeElement(vector<int>& nums, int val) {
			vector<int>::size_type i = 0;
			vector<int>::size_type j = nums.size() - 1;

			while (i < j+1) {
				if (nums[i] != val) {
					i++;
					continue;
				}

				if (nums[j] == val) {
					j--;
					continue;
				}

				nums[i++] = nums[j--];
			}

			return (int) j+1;
		}
};

File: java/RemoveElement.java
/*
 * Filename: RemoveElement.java
 * Author: Jun Xu
 * Email: junx@bu.edu
 * Created Time: Fri 13 Mar 2015 05:17:43 PM EDT
 */

public class Solution {
	public int removeElement(int[] A, int elem) {
		int ret = A.length;

		for (int i = 0; i < ret;) {
			if (A[i] == elem && A[ret-1] != elem) {
				A[i] = A[ret-1];
				ret--;
				i++;
			} else if (A[i] == elem && A[ret-1] == elem) {
				ret--;
			} else {
			    i++;
			}
		}

		return ret;
	}
}
